File: Icfpc2023/Domain.fs
namespace Icfpc2023

type Attendee = {
    X: double
    Y: double
    Tastes: double[]
}

type Pillar = {
    Center: PointD
    Radius: double
}

type Problem =
    {
        RoomWidth: double
        RoomHeight: double
        StageWidth: double
        StageHeight: double
        StageBottomLeft: PointD
        Musicians: int[]
        Attendees: Attendee[]
        Pillars: Pillar[]
    }
    /// Extension 2: Playing Together
    member this.IsTogetherExtensionActive =
        // Confirmed by the organizers: https://discord.com/channels/1118159165060292668/1126853058186444942/1127283248943353928
        this.Pillars.Length > 0

type IPartialSolution =
    abstract member GetPlacedMusicians: allMusicians: int[] -> seq<PointD * double * int>

type Solution =
    {
        Placements: PointD[]
        Volumes: double[]
    }
    interface IPartialSolution with
        member this.GetPlacedMusicians allMusicians =
            Seq.zip3 this.Placements this.Volumes allMusicians

module Solution =
    let defaultVolumes numOfMusicians = Array.replicate numOfMusicians 10.0

type Score = double
type SolverName = string

type SolutionMetadata = {
    Score: Score
    SolverName: SolverName
}

File: py/domain.py
from dataclasses import dataclass

import numpy as np


@dataclass
class Stage:
    x: float
    y: float
    w: float
    h: float


@dataclass
class Problem:
    room_width: float
    room_height: float
    stage_width: float
    stage_height: float
    stage_bottom_left: [float, float]
    musicians: [Musician]
    attendees: [Attendee]
    pillars: [Pillar]


@dataclass
class ProblemData:
    room_w: float
    room_h: float
    stage: Stage
    mus_instruments: np.ndarray
    att_places: np.ndarray
    att_tastes: np.ndarray
    pillar_center_radius: np.ndarray
    use_playing_together_ext: bool


@dataclass
class SolutionMetadata:
    score: float
    solver: str
